### ISSUE TITLE
Issue: Force closing MSFS-PyScriptManager will not close the script

### ISSUE THREAD

**Issue by leftos** [cgtrout/MSFS-PyScriptManager#12]
After shutting down the game, these lines spam the log:
```Error in get_time_to_future: get_simulator_datetime: Failed to retrieve simulator datetime: [WinError -1073741648] Windows Error 0xc00000b0```

and the status bar looks like this:
![Screenshot 2024-12-01 193343](https://github.com/user-attachments/assets/7496b968-13ff-478d-910a-8593a60e0fe8)

Obviously this isn't a big deal, at that point you likely want to terminate the script anyway, but it would be better if this was handled a bit more gracefully.

Perhaps catch that `winerror` and set `sim_connected` to `False`?

Haven't tested this (hoping to test it early this week), but this was roughly what I was thinking:
```    
except WindowsError.winerror as e:
    global sim_connected
    sim_connected = False
    print(f"Error in get_simulator_datetime: {e}")
    return datetime(0, 0, 0)  # Default to a zero datetime object
```

**Comment by cgtrout** [comment]
Let me know if these changes work for you.

How it appears now after exiting MSFS:
![image](https://github.com/user-attachments/assets/89270055-34f4-456a-b76f-e6f5bc6e9c7e)

**Comment by leftos** [comment]
You know how it nicely shows "Sim Not Running" when you first run the script if the sim isn't running? It would be nice if the script recovered gracefully in case the Simconnect connection is torn down (most likely because the user exited MSFS or it crashed), and went back to that state. Right now the NOT_CONNECTED text is better than the long Windows errors, but my overall hope was for the script to go back to a sim_connected False state so it can be reused without restarting it.

I haven't tested your latest change, but I also noticed that at the time of my last report, when the script went into that state, trying to right-click close any of my running scripts in the Script Manager UI would outright kill the whole application.

I'll have to test with your latest changes though and report back. Thanks!

**Comment by cgtrout** [comment]
I've been working on the issue of it more robustly handling restarts in a more user friendly way (no cryptic errors).

Commit a9aa09c - more notes on the changes are here, but with these changes the Sim can be restarted multiple times and the script will keep running.

**Comment by leftos** [comment]
So, although it all looks cleaner now after shutting down the sim (and I really appreciate that, thanks), the moment I try to close either of the script tabs in the script manager (again after having shut down the sim), the script manager appears to crash. The two scripts I'm running are still running in (now orphaned) pythonw.exe processes, but the script manager process is gone.

**Comment by cgtrout** [comment]
I've never encountered this behavior myself with any of the scripts myself, but I'll do some more testing to see if I can reproduce your issue.

Are you using the latest version of the "Launcher.py" script?  Also, when you have the issue, does the command line box for the c app show anything in the logs or does it also shut down automatically with no hints?

**Comment by leftos** [comment]
Yeah, I very recently did a pull of the head revision and just made some edits to the custom_status_bar script according to my preferences. I'm attaching my Launcher and Scripts folders if you want to do a diff.

When this happens I'm running custom_status_bar_leftos.py and virtual_pos_printer.py through an autolaunch group, if that makes any difference.

[MSFS-PyScriptManager.zip](https://github.com/user-attachments/files/18046499/MSFS-PyScriptManager.zip)

**Comment by cgtrout** [comment]
~~If you are willing, can you try to revert to a previous update for the launcher.py script to see if the issue still exists - such as the one bundled with the last release?  That would establish if the last set of updates to the launcher script are implicated on your issue.  With the latest launcher.py I added a performance monitor, and while that is intended to not at all change close behavior it is possible that is causing issues?~~

I'll have a chance to look into this further in a day or two but at the moment am away from the sim.

**Comment by cgtrout** [comment]
Actually looking at the zip you provided it looks like you already are running a previous version of the launcher script before I added that so that eliminates that theory.

**Comment by cgtrout** [comment]
To clarify: I believe that the launcher.py script is where the issue resides on this.

The design goal with the launcher app is that it closes any script regardless of what threading the script is doing.  The custom_status_bar.py did add a background_thread idea, so it's possible that this change has highlighted a new bug in the launcher.py script.

The MSFS-PyScriptManager.exe app calls the launcher.py script (which is what does all the script handling), and then the script output for that gets redirected to a console window which is minimized at open.

It looks like the following:
![image](https://github.com/user-attachments/assets/90f3f562-d5b7-4eb1-aef3-5f99db8abc88)

If possible can you show the output window when you get these issues.  Depending on how it crashes it's possible that window also closes on its own?  If it closes I may need to set up logging on the script and get you to reproduce it again so I can hopefully see an exception to determine the cause.

I'm not able to use anything with MSFS right now, but I have been trying to do some automated testing on the launcher which is testing multiple opens/closes of scripts with threads but as of yet I haven't been able to reproduce the issue you have reported.

Since the zip you gave me showed that you were running an older version of the launcher I'm wondering if it's possible other files are older files as well?  I might create a beta release or just send you a zip of my files to see if it still causes issues on your end.  That way we can try and get your environment exactly the same to ensure it isn't as simple as our files being out of sync.

**Comment by leftos** [comment]
Yeah, my files should be in sync with what's posted on GitHub, but I can do a full pull again and replace everything, since my custom script has a different name, if that would help.

And yes, I'm aware of the launcher window, I'll keep an eye on its output after I close MSFS 2020 next.

I also might not get to test and get back to you for a couple of days, but I've been tracking this thread, and I'll follow your suggestions and report back when possible.

**Comment by leftos** [comment]
Without changing any of my files, the console window of the pyscriptmanager had closed after I had closed the sim. So It appears that perhaps the launcher itself is crashing? I can try again with the very latest from the repo.

**Comment by leftos** [comment]
I tried to compile from the latest sources, and it doesn't appear that the launcher exe can launch the launcher.py:
```
MSFS-PyScriptManager: Loader exe
-------------------------------------------------------------------------------------------

Launching Python script: .\Launcher\LauncherScript\launcher.py
Error: CreateProcess failed. Error code: 2
Press any key to exit...
```

EDIT: Never mind, I'm dumb, I didn't download and install WinPython, only TCC.

**Comment by leftos** [comment]
It would be good to include all the required modules in your readme.md for winpython as well, like ttkthemes, psutil, keyboard, and so on, since those aren't included with the WinPython release you link to.

**Comment by leftos** [comment]
So, I booted the script manager exe from the terminal so I could catch the output when it crashed, but there was nothing of interest.

```
PS C:\FS_Addons\MSFS-PyScriptManager> .\MSFS-PyScriptManager.exe
MSFS-PyScriptManager: Loader exe
-------------------------------------------------------------------------------------------

Launching Python script: .\Launcher\LauncherScript\launcher.py
Reading Python script output...

Keep this window open to monitor launcher.py output, otherwise it is safe to close.
-------------------------------------------------------------------------------------------

Autoplay: Loading script group from C:\FS_Addons\MSFS-PyScriptManager\Scripts\_autoplay.script_group
run_script: C:\FS_Addons\MSFS-PyScriptManager\Scripts\custom_status_bar_leftos.py tab_id: 1
run_script: C:\FS_Addons\MSFS-PyScriptManager\Scripts\virtual_pos_printer.py tab_id: 2
PS C:\FS_Addons\MSFS-PyScriptManager>
```

**Comment by cgtrout** [comment]
Thank you for everything you have tried here.

<strike>
Are you able to zip up everything you have (except for the WinPython dir) so I can test everything you have?  Alternatively I can zip up my dev dir for you to test.

I want to establish first if everything occurs exactly the same if I try your dir or you try mine.  That would eliminate the possibility that we simply have slightly different files.

If that doesn't work then I will have to try and determine why it is not closing properly for you with your configuration.

My current dev dir if you want to try it:
https://onedrive.live.com/?redeem=aHR0cHM6Ly8xZHJ2Lm1zL2YvYy80NDdhZDM2OTNhNzA0OGY0L0V2UkljRHBwMDNvZ2dFU3R6d0FBQUFBQklSbDQzTy1Fcmx5cXBJNnd5SkR5Q2c_ZT0wdGZZWGs&id=447AD3693A7048F4%2153165&cid=447AD3693A7048F4
</strike>

I've tried again with mine - went into a flight, exited out of the sim, and after it closed I was able to close each script with no issues.  So either we have different configuration on MSFS-PyScriptManager, or there is something specific to your hardware/OS that causes the script to fail.

Edit: After consideration, I do not think you need to try my download.  I've tested multiple versions of launcher.py and it is hard to imagine now that there is a specific file you could have that would explain this.

note that when I run and close tabs it will give logs for the closing of the tabs- if those do not show on yours when you try to close that might be the best hint we have yet:
![image](https://github.com/user-attachments/assets/9c561c5f-18b8-4701-8c66-8fbe8877c9be)

**Comment by cgtrout** [comment]
Last night I was making some changes to launcher.py
https://github.com/cgtrout/MSFS-PyScriptManager/blob/ce36cfc1c01ac01585cb25f06c115929744c7d25/Launcher/LauncherScript/Launcher.py

This is from another branch - you can try this to see if it runs any differently.  Since I don't yet know why the issue is occuring theres a good chance this won't make any difference for you.  But I've tested this new version and the latest on the main branch and both work for me.

**Comment by cgtrout** [comment]
Based on what you've told me this occurs exactly when you try to right click to close a tab?  Also as you showed, it seems the close handling for the tabs maybe is never even called?

I've made one logging change to the right click handler on Launcher.py (that closes the tabs).  

https://github.com/cgtrout/MSFS-PyScriptManager/commit/4fb6bdb7aa4d18790d3a624202899c9ad3865aa1

```
def on_tab_right_click(self, event):
        """Handle right-click event on notebook tabs for direct tab closing."""
        try:
            # Get the index of the clicked tab
            print("DEBUG: on_tab_right_click: enter")
            clicked_tab_index = self.notebook.index(f"@{event.x},{event.y}")
            print(f"Right-clicked on tab with index: {clicked_tab_index}")
            self.close_tab_by_index(clicked_tab_index)  # Directly close the tab
        except TclError as e:
            print(f"Error: on_tab_right_click error - Details: {e}")
```

This will give logging that first this function is actually called, and then secondly if it fails on a TclError it will print something.  This won't solve the issue but may give some more hints on if this actually gets called.

**Comment by cgtrout** [comment]
https://github.com/cgtrout/MSFS-PyScriptManager/blob/fix/shutdown-issues/Launcher/LauncherScript/Launcher.py

I've made several changes to the launcher script.  One of my central theories on this is that your issue is a multithreading issue of some kind on the launcher script.  Maybe we have different CPUs, OS's etc that change the thread scheduling.  This could explain why I have no issues and you do.

I have went through the entire script and tried to identify problem areas related to multithreading handling.  The launcher now makes more judicious use of locks to ensure that associated tab states can't be modified incorrectly by multiple threads.  There were definately some problem areas which could have contributed to the problems you have been seeing.

Try this launcher script with the app to see if it resolves your issue.  Unfortunately I have not found a way myself to reproduce your issue but hopefully these last changes to Launcher.py will help move things in the correct direction.  If it doesn't work, I might work towards using more modern multithreading constructs on the script as it perhaps relies too heavilly on old locking constructs.

**Comment by leftos** [comment]
I'm... pretty sure I've been sort of wasting your time on this one, but maybe not. 

I just realized that I've been using Addon Linker to launch the PyScriptManager exe as part of booting up MSFS 2020, but I also had the setting checked to automatically close it on exit. Unfortunately, on exit it would only close the exe, not the process tree it spawned, so by the .exe now being closed, trying to close a tab crashes the launcher.py script.

Now, that said, the .exe's output says that unless you want to monitor launcher.py, you can close the .exe, but if you do that, again, simply trying to right-click on a tab to close it seems to crash launcher.py.

So either the output text of the .exe should be updated to remove that suggestion, or launcher.py should be updated not to be dependent on the .exe running for something like a close tab operation, imho.

**Comment by cgtrout** [comment]
Thanks for the updated info!  

I am now able to reproduce the issue myself.  I had tested that behavior a couple months ago and it didn't cause any issues then which caused me to (incorrectly) rule it out for your problem.

I have created a fix for this in the c app:
https://github.com/cgtrout/MSFS-PyScriptManager/commit/a382fd8fdd9338dca79b9b8ea8650203c1597cb6

The cause is that the 'piping' that redirected the output of the launcher script would ~~not~~ close down when the c app was closed which would cause the launcher script to crash when it tried to write any ouput to its std (output).   

To fix, replace this "Launcher.c" c file with the updated version in the link, and then run "Build.bat".  This will compile a new exe.

I'll get this rolled into the official version at next release, but I'd appreciate if you can test and confirm that this fixes things for you.  

Thanks again for the feedback - glad we are converging towards a solution.

**Comment by cgtrout** [comment]
Maybe the exe behavior should change - would it be preferrable if the script launcher is closed as well when the exe is closed - that would make it more automateable to close it?

**Comment by leftos** [comment]
> Maybe the exe behavior should change - would it be preferrable if the script launcher is closed as well when the exe is closed - that would make it more automateable to close it?

Honestly, I'm just one user, so take my preference with a grain of salt (and the moment I typed preference I realized this could be a command line argument to the launcher exe), but yes, it would be nice if I can just close the exe and have it also close the launcher itself, and by extension, all running scripts.

**Comment by leftos** [comment]
So the new launcher.c fixed the issue, however I didn't see any output in it about the tabs opening in the launcher, like I used to see before.

**Comment by cgtrout** [comment]
> So the new launcher.c fixed the issue, however I didn't see any output in it about the tabs opening in the launcher, like I used to see before.

Thanks for pointing that out.  Now fixed:
https://github.com/cgtrout/MSFS-PyScriptManager/commit/19e50c355955810d2c46b578a70bc242469f4f3c

**Comment by leftos** [comment]
Output confirmed fixed.

**Comment by cgtrout** [comment]
Note that the shutdown idea (where closing the c app causes script to also close) is implemented in the current dev branch - https://github.com/cgtrout/MSFS-PyScriptManager/tree/new-version-update

I was hoping to have a new release ready by now, but it will probably be a week or two until it's ready.

Specific commit with the change is: https://github.com/cgtrout/MSFS-PyScriptManager/commit/f05060974afbdc37d9caa693599638cdad2b7fb2

Note that this requires a 'proper' shutdown.  I'm not sure what method Addon Linker uses for this to close the app, but if it force closes this new idea won't work.  

If you try this out, let me know if it works with AddonLinkers automated shutdown.  If this idea doesn't work I can try a slightly different idea.

**Comment by cgtrout** [comment]
I will close this issue for now, but feel free to reply if you have issues related to this.

**Comment by leftos** [comment]
Hmmm. It doesn't look like whatever signal MSFS Addon Linker is sending is able to terminate the process. And of course, if you opt to force kill it, like you said, that doesn't get handled properly and doesn't actually gracefully shut down Launcher.py.

I figured all it's doing is a taskkill, and decided to try it out myself...
```
C:\Users\Leftos>taskkill /im MSFS-PyScriptManager.exe
ERROR: The process "MSFS-PyScriptManager.exe" with PID 7584 could not be terminated.
Reason: This process can only be terminated forcefully (with /F option).
```

So that checks out.

**Comment by cgtrout** [comment]
I had one other idea on this to have a repeating signal/hearbeat to monitor - when that ends it can be closed.

This I will place at a fairly low priority for now.

**Comment by leftos** [comment]
Yeah, low priority sounds right.

**Comment by cgtrout** [comment]
The "heartbeat-shutdown" functionality (as discussed earlier ) has been merged in with new-version-update(https://github.com/cgtrout/MSFS-PyScriptManager/pull/45)

It adds a pipe command, "HEARTBEAT", that if not detected after 5 seconds will force a clean shutdown of the Launcher.py script. 
 This has been tested with the "taskkill /IM MSFS-PyScriptManager.exe /F" command and is confirmed to work properly with my own testing.

Note that since this changes the Launcher.c behavior you will need to **recompile with Build.bat** for this to work properly.

Let me know if this resolves your issue - thanks!

**Comment by leftos** [comment]
Confirmed resolved. Thanks so much.
